Change MCMC kernels to not depend on trace

neerajprad opened the pull request:
As discussed in #1845, this renames `TraceKernel` to `MCMCKernel` and changes the interface so that the kernels take in dicts instead of model traces. 

I have tried to make minimal changes. ~but since the current interface is very coupled with the `model`, I had to make slightly more changes than I had planned~. Existing tests work as expected.

neerajprad commented:
@fehiepsi - Feel free to commit to this PR directly, if there is anything you would like to add. I think this should be independent of your changes.

neerajprad commented:
Hmm..this change is not compatible with `pyro.param` statements in the model captured by the failing test. While we don't need `pyro.param` for MCMC, our existing implementation that uses `prototype_trace` handles this without any issues, whereas this proposed change doesn't. I'll think about a fix for this.

fehiepsi commented on pyro/infer/mcmc/mcmc_kernel.py:
```diff
@@ -42,34 +42,34 @@ def end_warmup(self):
+    def sample(self, params):
         """
-        Samples a trace from the approximate posterior distribution, when given an existing trace.
+        Samples parameters from the posterior distribution, when given an existing trace.
```
typo: when given existing *parameters*.

fehiepsi commented on pyro/infer/mcmc/mcmc_kernel.py:
```diff
@@ -42,34 +42,34 @@ def end_warmup(self):
+        Samples parameters from the posterior distribution, when given an existing trace.
 
-        :param trace: Current execution trace.
+        :param trace: Current parameter values.
```
typo: `:param params:`

fehiepsi commented on pyro/infer/mcmc/mcmc.py:
```diff
@@ -247,6 +248,7 @@ class MCMC(TracePosterior):
     """
     def __init__(self, kernel, num_samples, warmup_steps=None,
                  num_chains=1, mp_context=None, disable_progbar=False):
+        self.kernel = kernel
```
We need to revise docs for `kernel` too.

fehiepsi commented on pyro/infer/mcmc/mcmc.py:
```diff
@@ -264,9 +266,21 @@ def __init__(self, kernel, num_samples, warmup_steps=None,
         super(MCMC, self).__init__(num_chains=num_chains)
 
+    # TODO: Refactor so that the MCMC class directly has access to the
+    # trace generator and transforms needed to do this wrapping.
```
Could you also add a TODO for mentioning about the params from kernel lie in unconstrained space?

neerajprad commented:
@fehiepsi - I have addressed your comments, but I realized that we will need to address the TODO to make the multiprocessing tests pass. The issue is that we cannot access `prototype_trace` or `transforms` from `self.kernel` in `MCMC` since they are set in a different process with python multiprocessing. 

This will require a refactoring so that we can pass model to `MCMC` directly and call `initialize_model` to get `transforms`, `prototype_trace`, `initial_params` within the `MCMC` class. That makes it much harder to have a backwards compatible change though. I am thinking that it might be worth breaking backwards compatibility and clean up all the interface issues for once. 

For the time being, please feel free to pull in these changes into your PR, or we can merge your PR first in case that's ready.

fehiepsi commented:
Thanks @neerajprad ! I'll pull these changes to my PR and think about what we can do to address multiprocessing issue.

fehiepsi commented:
@neerajprad How about move `trace_wrap` logic to workers? There it is safe to get access to prototype trace and transform_fn.

neerajprad commented:
> @neerajprad How about move trace_wrap logic to workers? There it is safe to get access to prototype trace and transform_fn.

That's a good idea, let me change this and report back.

neerajprad commented:
@fehiepsi - your suggestion worked great. I think this should be good to merge once tests pass, pending further comments.

fehiepsi commented:
Yeah, I also have an almost working version in my branch. Currently, tests only fail for some edge cases such as empty param or adapt_step_size=False.
